src/client/Microsoft.Identity.Client/Platforms/Features/SLC/KeyMaterialManager.cs

```diff
@@ -0,0 +1,206 @@
+                }
+            }
+
+            _isInitialized = false;
+            return null;
+        }
+
+        /// <summary>
+        /// Determines if a given X.509 certificate needs rotation based on a percentage threshold.
+        /// </summary>
+        /// <param name="certificate">The X.509 certificate to evaluate.</param>
+        /// <param name="rotationPercentageThreshold">The threshold percentage for considering rotation (default is 70%).</param>
+        /// <returns>
+        /// True if the certificate needs rotation, false otherwise.
+        /// </returns>
+        public static bool CertificateNeedsRotation(X509Certificate2 certificate, double rotationPercentageThreshold = 70)
+        {
+            DateTime now = DateTime.UtcNow;
+
+            // Calculate the total duration of the certificate's validity
+            TimeSpan certificateLifetime = certificate.NotAfter - certificate.NotBefore;
+
+            // Calculate how much time has passed since the certificate's issuance
+            TimeSpan elapsedTime = now - certificate.NotBefore;
+
+            // Calculate the current percentage of the certificate's lifetime that has passed
+            double percentageElapsed = (elapsedTime.TotalMilliseconds / certificateLifetime.TotalMilliseconds) * 100.0;
+
+            // Check if the percentage elapsed exceeds the rotation threshold
+            return percentageElapsed >= rotationPercentageThreshold;
+        }
+
+        /// <summary>
+        /// Creates a binding certificate with a the key material for use in Managed Identity scenarios.
+        /// </summary>
+        /// <param name="eCDsaKey">The key used for creating the certificate.</param>
+        /// <returns>The created binding certificate.</returns>
+        private X509Certificate2 CreateBindingCertificate(ECDsa eCDsaKey)
+        {
+            try
+            {
+                lock (_keyInfoLock) // Lock to ensure thread safety
+                {
+                    _logger.Verbose(() => "[Managed Identity] Creating binding certificate " +
+                    "with CNG key for credential endpoint.");
+
+                    // Create a certificate request
+                    CertificateRequest request = CreateCertificateRequest(CertSubjectname, eCDsaKey);
+
+                    // Create a self-signed X.509 certificate
+                    DateTimeOffset startDate = DateTimeOffset.UtcNow;
+                    DateTimeOffset endDate = startDate.AddYears(5); //expiry 
+
+                    //Create the self signed cert
+                    X509Certificate2 selfSigned = request.CreateSelfSigned(startDate, endDate);
+
+                    if (!selfSigned.HasPrivateKey)
+                    {
+                        _logger.Error("[Managed Identity] The Certificate is missing the private key.");
+                        throw new InvalidOperationException("The MTLS Certificate must include a private key.");
```

Comment:
if you go the route of exceptions, throw `MsalClientException(cert_provision_error, "Cannot create a self signed certificate")`.

Remember that app developers have no idea that you use MTLS and this error message if for them.

Reply:
Actually just throw        `throw new MsalClientException(MsalError.CertificateCreationFailed,
                    $"Failed to create Managed Identity binding certificate. Error : {ex.Message}");`